mlir/lib/Dialect/MIOpen/Transforms/ConvToGemmLowering.cpp

```diff
@@ -1582,10 +1583,85 @@ template struct Conv2DRewritePattern<Conv2DOp>;
 template struct Conv2DRewritePattern<Conv2DBwdDataOp>;
 template struct Conv2DRewritePattern<Conv2DBwdWeightOp>;
 
+//===- MITPRewritePattern -------------------------------------------------===//
+//===-  ------------------------------------------------===//
+template <typename T> struct MITPRewritePattern : public OpRewritePattern<T> {
+  using OpRewritePattern<T>::OpRewritePattern;
+
+  miopen::TransformOp makeTranspose(PatternRewriter &b, Value inp,
+                                    const AffineMapAttr &inMap,
+                                    const AffineMapAttr &outMap) const {
+    auto inpIdxMap = inMap.getAffineMap();
+    auto outpIdxMap = outMap.getAffineMap();
+    auto loc = inp.getLoc();
+    auto inpType = inp.getType().template cast<MemRefType>();
+    auto inpShape = inpType.getShape();
+
+    SmallVector<uint32_t, 8> endDims;
+    SmallVector<uint32_t, 8> startDims;
+    for (uint32_t i = 0, e = inpShape.size(); i < e; ++i) {
+      startDims.push_back(i);
+      auto inMapped = inpIdxMap.getDimPosition(i);
+      endDims.push_back(outpIdxMap.getDimPosition(inMapped));
+    }
+    miopen::BottomUpTMBuilder transform(b, inpShape, loc);
+    transform.passThrough(endDims, startDims);
+    auto tfOp = b.create<miopen::TransformOp>(loc, inp, transform.get(),
+                                              inpType.getMemorySpaceAsInt());
+    return tfOp;
+  }
+
+  LogicalResult matchAndRewrite(T laGeneric,
+                                PatternRewriter &b) const override {
+    LogicalResult fail = failure();
+    auto loc = laGeneric.getLoc();
+
+    // 0. Test compatibility
+    // 0.0. Only fully parallel for now
+    for (auto itr : laGeneric.iterator_types()) {
```

Comment:
If there's a range, there's a method on `ArrayAttr` that does the casting for you.

Heck, `.getAsValueRange<StringAttr>()` should let you write this as `for (StringRef iter : ...)`